central/externalbackups/plugins/s3compatible/s3compatible.go

```diff
@@ -0,0 +1,238 @@
+	}
+	if cfg.GetRegion() == "" {
+		errorList.AddString("Region must be specified")
+	}
+
+	return errorList.ToError()
+}
+
+func validateEndpoint(endpoint string) error {
+	if _, err := url.Parse(fmt.Sprintf("https://%s", urlfmt.TrimHTTPPrefixes(endpoint))); err != nil {
+		return errors.Wrapf(err, "invalid URL %s", endpoint)
+	}
+	return nil
+}
+
+func newS3Compatible(integration *storage.ExternalBackup) (*s3Compatible, error) {
+	cfg, ok := integration.GetConfig().(*storage.ExternalBackup_S3Compatible)
+	if !ok {
+		return nil, errors.New("S3 Compatible configuration required")
+	}
+	if err := validate(cfg.S3Compatible); err != nil {
+		return nil, err
+	}
+
+	opts := []func(*config.LoadOptions) error{
+		config.WithRegion(cfg.S3Compatible.GetRegion()),
+		config.WithHTTPClient(&http.Client{Transport: proxy.RoundTripper()}),
+	}
+
+	if cfg.S3Compatible.GetUseIam() {
+		opts = append(opts, config.WithCredentialsProvider(
+			credentials.NewStaticCredentialsProvider(
+				cfg.S3Compatible.GetAccessKeyId(),
+				cfg.S3Compatible.GetSecretAccessKey(), "",
+			),
+		))
+	}
+
+	ctx, cancelFn := context.WithTimeout(context.Background(), initialConfigurationMaxTimeout)
+	defer cancelFn()
+	awsConfig, err := config.LoadDefaultConfig(ctx, opts...)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to load the aws config")
+	}
+
+	var clientOpts []func(*s3.Options)
+	if cfg.S3Compatible.GetUsePathStyle() {
+		clientOpts = append(clientOpts, func(o *s3.Options) {
+			o.UsePathStyle = true
+		})
+	}
+
+	if endpoint := cfg.S3Compatible.GetEndpoint(); endpoint != "" {
+		if err := validateEndpoint(endpoint); err != nil {
+			return nil, err
+		}
+		clientOpts = append(clientOpts, func(o *s3.Options) {
+			o.BaseEndpoint = aws.String(endpoint)
+		})
+	}
+
+	client := s3.NewFromConfig(awsConfig, clientOpts...)
+	return &s3Compatible{
+		integration: integration,
+		client:      client,
+		uploader:    manager.NewUploader(client),
+	}, nil
+}
+
+func (s *s3Compatible) Backup(reader io.ReadCloser) error {
+	defer func() {
+		if err := reader.Close(); err != nil {
+			log.Errorf("closing reader: %+v", err)
+		}
+	}()
+
+	log.Info("Starting S3 Compatible Backup")
+	ctx, cancelFn := context.WithTimeout(context.Background(), backupMaxTimeout)
+	defer cancelFn()
+	formattedTime := time.Now().Format("2006-01-02T15:04:05")
+	key := fmt.Sprintf("backup_%s.zip", formattedTime)
+	formattedKey := s.prefixKey(key)
+	if _, err := s.uploader.Upload(ctx, &s3.PutObjectInput{
+		Bucket: aws.String(s.integration.GetS3Compatible().GetBucket()),
+		Key:    aws.String(formattedKey),
+		Body:   reader,
+	}); err != nil {
+		return s.createError(fmt.Sprintf("error creating backup in bucket %q with key %q",
+			s.integration.GetS3Compatible().GetBucket(), formattedKey), err)
+	}
+	log.Info("Successfully backed up to S3")
+	return s.pruneBackupsIfNecessary(ctx)
+}
+
+func (s *s3Compatible) createError(msg string, err error) error {
```

Comment:
same here

Reply:
There is a bug in the current state of this function because it uses `awserr` from the `aws-sdk-go`, not the `v2` (copy-paste mistake 😅 ). I suspect that once it uses the `v2` equivalent might have the same problem as `sortS3Objects`.